src/procedures/__tests__/ModifyTieredStoData.ts

```diff
@@ -0,0 +1,780 @@
+      expect(
+        addTransactionSpy
+          .getCall(1)
+          .calledWith(tieredSto_3_1_0_Mock.getMockInstance().modifyFunding)
+      ).toEqual(true);
+      expect(addTransactionSpy.getCall(1).lastArg.tag).toEqual(PolyTransactionTag.ModifyFunding);
+      expect(
+        addTransactionSpy
+          .getCall(2)
+          .calledWith(tieredSto_3_1_0_Mock.getMockInstance().modifyOracles)
+      ).toEqual(true);
+      expect(addTransactionSpy.getCall(2).lastArg.tag).toEqual(PolyTransactionTag.ModifyOracles);
+      expect(
+        addTransactionSpy.getCall(3).calledWith(tieredSto_3_1_0_Mock.getMockInstance().modifyTiers)
+      ).toEqual(true);
+      expect(addTransactionSpy.getCall(3).lastArg.tag).toEqual(PolyTransactionTag.ModifyTiers);
+      expect(
+        addTransactionSpy.getCall(4).calledWith(tieredSto_3_1_0_Mock.getMockInstance().modifyLimits)
+      ).toEqual(true);
+      expect(addTransactionSpy.getCall(4).lastArg.tag).toEqual(PolyTransactionTag.ModifyLimits);
+      expect(
+        addTransactionSpy
+          .getCall(5)
+          .calledWith(tieredSto_3_1_0_Mock.getMockInstance().modifyAddresses)
+      ).toEqual(true);
+      expect(addTransactionSpy.getCall(5).lastArg.tag).toEqual(PolyTransactionTag.ModifyAddresses);
+      expect(addTransactionSpy.callCount).toEqual(6);
+    });
+
+    test('should throw an error if custom currencies are used with 3_0_0 wrapper', async () => {
+      target = new ModifyTieredStoData(
+        {
+          ...tieredParams,
+          customCurrency,
+        },
+        contextMock.getMockInstance()
+      );
+      moduleWrapperFactoryMock.mock('getModuleInstance', tieredSto_3_0_0_Mock.getMockInstance());
+      await expect(target.prepareTransactions()).rejects.toThrow(
+        new PolymathError({
+          code: ErrorCode.InvalidAddress,
+          message: `Custom currency not supported in Tiered STO v3.0`,
+        })
+      );
+    });
+
+    test('should throw an error if the tiered sto has not yet been launched or is archived', async () => {
+      moduleWrapperFactoryMock.mock('getModuleInstance', Promise.resolve(undefined));
+
+      await expect(target.prepareTransactions()).rejects.toThrow(
+        new PolymathError({
+          code: ErrorCode.InvalidAddress,
+          message: `STO ${tieredParams.stoAddress} is either archived or hasn't been launched`,
+        })
+      );
+    });
+
+    test('should throw if the tiered sto data params do not present valid modifications, there are no transactions to add to queue', async () => {
+      await expect(target.prepareTransactions()).rejects.toThrow(
+        new PolymathError({
+          code: ErrorCode.InvalidAddress,
+          message: `Modify STO data failed: nothing to modify`,
+        })
+      );
+    });
+
+    test('should throw if the tiered sto data params do not present valid modifications, there are no transactions to add to queue and no fundraise currencies fetched from the entity', async () => {
```

Comment:
What is the purpose of this test? Isn't the "no transactions to add" case already covered in the previous test?

Reply:
This test was weird yes. I added it to complete 100% coverage. Its possible that this could be in a different form of test. The conditions of this

```
  if (!currencies) {
      currencies = [];

      if (isRaisedInETH) {
        currencies.push(Currency.ETH);
      }

      if (isRaisedInPOLY) {
        currencies.push(Currency.POLY);
      }

      if (isRaisedInSC) {
        currencies.push(Currency.StableCoin);
      }
    }
```
This test basically tests the case where !currencies. The test on line 702 is the case where !currencies and isRaisedInETH, isRaisedInPoly, isRaisedInSc are **valid**. This test is the case where !currencies and !isRaisedInEth, !isRaisedInPoly, !isRaisedInSc. In both tests, it is a case with nothing to modify.

Reply:
I see your point. In that case, those code branches should be covered inside the same test case. What you would do is something like:

1) Run the test as normal
2)
- Mock fetch to return a Tiered STO where currencies is just ETH
- Run prepareTransactions again and check for the same error
3)
- Mock fetch to rerturn a Tiered STO where currencies is just POLY
- Run prepareTransactions again and check for the same error

The reason behind simply not having the mocked STO's currencies be an empty array is because that is an unrealistic case that cannot happen

The idea behind this is that a test case should test for a specific behavior (in a way that can be seen by the user), not implementation details. A test case definition should be concerned with the *what*, not the *how*. So if you're ever in a situation where you're having to write a separate test case just to test for a specific code branch or condition, then you should rethink your approach